content/blog/2019-11-13-llvm-profiling/index.md

```diff
@@ -0,0 +1,336 @@
+      
+    vector<string> atomicCounter; //keep global variable names for profiling. e.g. instr counter
+  };
+```
+
+
+
+## Implementation
+
+### Basic Execution Information
+
+We first create global variable `llvmInstrAtomicCounter` and `basicBlockAtomicCounter` with [GlobalVariable](<https://llvm.org/doxygen/classllvm_1_1GlobalVariable.html#a3ef813d6bda7e49e31cb6bf239c4e264>) constructor.
+
+```c++
+new GlobalVariable(M, I64Ty, false, GlobalValue::CommonLinkage, ConstantInt::get(I64Ty, 0), atomicCounter[i]);
+```
+
+ Then in each basic block we obtain pointer to the global variable names with [getOrInsertGlobal](<https://llvm.org/doxygen/classllvm_1_1Module.html#abd8f7242df6ecb10f429c4d39403c334>) method. After getting the instruction number in each block, we create atomic addition with [AtomicRMWInst](<https://llvm.org/doxygen/classllvm_1_1AtomicRMWInst.html#abf7e0649c7f272cc49165e579be010a5>) constructor. The complete code looks like this:
+
+```c++
+
+void SkeletonPass::createInstr(BasicBlock &bb, Constant *counter_ptr, int num){
+    // create atomic addition instruction
+    if(num){
+        new AtomicRMWInst(AtomicRMWInst::Add,
+                      counter_ptr, // pointer to global variable
+                      ConstantInt::get(Type::getInt64Ty(bb.getContext()), num), //create integer with value num
+                      AtomicOrdering::SequentiallyConsistent, //operations may not be reordered
+                      SyncScope::System, // synchronize to all threads
+                      bb.getTerminator());//insert right before block terminator
+    }
+}
+
+bool SkeletonPass::runOnBasicBlock(BasicBlock &bb, Module &M)
+{
+    // get the global variable for atomic counter
+    Type *I64Ty = Type::getInt64Ty(M.getContext());
+    Constant *instrCounter = M.getOrInsertGlobal("llvmInstrAtomicCounter", I64Ty);
+    assert(instrCounter && "Could not declare or find llvmInstrAtomicCounter global");
+    Constant *bbCounter = M.getOrInsertGlobal("basicBlockAtomicCounter", I64Ty);
+    assert(bbCounter && "Could not declare or find basicBlockAtomicCounter global");
+    // get instruction number and basic block number.
+    int basic_block = 1;
+    int instr = bb.size();
+    // create atomic addition instruction
+    createInstr(bb, bbCounter, basic_block);
+    createInstr(bb, instrCounter,instr);
+    
+    return true;
+}
+
+```
+
+Finally, we print the profiling results by first create a function `printf` by specifying input argument and output argument types. Then transverse the `main` function. Right before `return` of the last block in `main`, create string variable corresponding to variables we are tracing and get the pointer to that string variable. Also, get the number variable by first get the pointer pointing to the global variable and then create a load instruction to get the actual pointer to the value. Finally, we can call printf function with input arguments.
```

Comment:
I'm a little lost about the high-level purpose of this work. Why are there string variables involved? Why isn't the template argument to `printf` just a constant string?

The specific mechanics of exactly how you generated the instructions are less important than making it clear what the instrumented program does and why.